auth/basic_auth.go

```diff
@@ -15,12 +15,38 @@ func DecorateWithBasicAuth(next http.HandlerFunc, credentials *BasicAuthCredenti
 		w.Header().Set("WWW-Authenticate", `Basic realm="Restricted"`)
 
 		if !ok || !(credentials.Password == password && user == credentials.User) {
+			Unauthorized(w, r)
+		}
+
+		next.ServeHTTP(w, r)
+	}
+}
+
+// Support for Multiple User Basic Auth Providers
+func DecorateWithMultiUserBasicAuth(next http.HandlerFunc, creds []BasicAuthCredentials) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
 
-			w.WriteHeader(http.StatusUnauthorized)
-			w.Write([]byte("invalid credentials"))
-			return
+		ok := false
+		user, password, _ := r.BasicAuth()
+		w.Header().Set("WWW-Authenticate", `Basic realm="Restricted"`)
+
+		for _, v := range creds {
+			if v.User == user && v.Password == password {
```

Comment:
I see the logic overlapping, so why don't we just call `DecorateWithBasicAuth` here per single `BasicAuthCredentials`. The only comment is that really.

Reply:
@martindekov that would spawn multiple auth windows for each user I beleive. We could potentially extract it out to another function to spawn the auth though

Reply:
Martin is right, but I think we want the inverse relationship, `DecorateWithBasicAuth` should delegate to `DecorateWithMultiUserBasicAuth`


In the current implementation this would look like
```go
func DecorateWithBasicAuth(next http.HandlerFunc, credentials *BasicAuthCredentials) http.HandlerFunc {
     return DecorateWithMultiUserBasicAuth(next, []BasicAuthCredentials{*credentials})
}
```

But this also highlights a mismatch in design between the two. `DecorateWithMultiUserBasicAuth` should accept points to `BasicAuthCredentials`, not the instances, if it is going to match the previous implementation.  Perhaps @alexellis can comment on why the pointer is needed here.   I am not sure the pointers are needed _and_ there is no check for `nil` in the previous implementation, so there is a possible bug there.

In combination with my previous comment, this would look like
```go
func DecorateWithBasicAuth(next http.HandlerFunc, credentials BasicAuthCredentials) http.HandlerFunc {
     return DecorateWithMultiUserBasicAuth(next, map[string]BasicAuthCredentials{credentials.User: credentials})
}
```